Path exploration goodies

karkhaz opened the pull request:
Most of this daunting pull request is tests, I promise. :innocent:

- Implement @martin-cs's suggestion that control of which path to continue executing be moved out of `goto_symex/symex_goto.cpp` entirely. My previous implementation always saved one path on the queue and continued executing the other, meaning that _every_ symex run ended at the end of a path. We now save _both_ paths and halt symbolic execution, returning control to the caller, who can choose which path to execute (either of the two paths that were just saved, or even any of the paths that were saved earlier). This implies that a symex run shouldn't always be followed by a safety check, since we may not have finished executing a path; thus, I added `safety_checkert::resultt::SUSPENDED` to indicate that symex was suspended half-way through and the resulting equation has not been safety-checked.

- The path workqueue now has a custom data type instead of being a `std::list`, with `push()`, `peek()`, and `pop()` methods. The data type is subclassable for implementing different strategies. `peek()` is intentionally not marked `const`, so that the data type can lazily search for the next path to return only when `peek()` is called rather than modifying its state every time `push()` is called; I anticipate that future strategies might be time-consuming.

- There's a data structure called `path_strategy_choosert` which is both a factory for path exploration strategies, and also holds their stringified names and help texts all in the same place, so that there's a uniform way of setting the path exploration strategy from all the frontends without duplication.

- Two strategies are implemented: FIFO and progressive FIFO. The paths that FIFO returns will always alternate between the next instruction of a goto, and the target of that same goto, followed by the next and target of another goto, etc. Progressive FIFO will always return the jump targets of all gotos that have been pushed, and only start returning next instructions when the jump targets have all been executed---the idea being to prefer making forward progress rather than dwelling inside loops.

- Tests for all of this, courtesy of the Department of Overengineered Solutions. In order to test the functional correctness of strategies, we not only need to test the verification result, but also the order in which paths are popped. This means multi-line regexes in the test suite, but even for short test cases these regexes are hundreds of bytes long---tedious to write and impossible to understand. I therefore wrote an interpreter for a human-friendly language representing the pushes and pops of a path queue. Test writers should write what paths they expect to see in what order in the comments section of a `test.in` file, from which a `test.desc` file gets generated by `scripts/make_descs.py`. For example, writing this `test.in` file:

```
CORE
main.c
--unwind 2
activate-multi-line-match
SIGNAL=0
EXIT=10
--
^warning: ignoring
--
This strategy differs from 'fifo' because the failed path is the second
one, not the last one.

Test that the following happens in order:
execute line 6
save next 7
save jump 9

// We skip the loop first with this strategy. This path is infeasible,
// since x is 1 so we must have entered the loop; so the solver is happy.
any lines
execute line 6
resume jump 9
execute line 9
any lines
path is successful

// We now enter the loop and execute it; we then save "entering the loop
// twice" and "bailing out after the first spin".
any lines
execute line 6
resume next 7
execute line 7
any lines
save next 7
save jump 9

// Since the just saved "bailing out after the first spin" is a jump,
// execute that path. That path makes the assertion fail, since we
// decremented x from 1 to 0.
any lines
execute line 6
resume jump 9
execute line 9
any lines
path is unsuccessful

// Finally, execute "entering the loop twice". This path is infeasible
// because we cannot actually have entered the loop twice if x was only
// 1 to begin with, so the solver is happy.
any lines
execute line 6
resume next 7
execute line 7
any lines
path is successful
end
```
generates this `test.desc:`
```
CORE
main.c
--unwind 2
activate-multi-line-match
SIGNAL=0
EXIT=10
BMC at file[^\n]+line 6 function \w+\nSaving next instruction 'file[^\n]+line 7 function \w+'\nSaving jump target 'file[^\n]+line 9 function \w+'(.|\n)+BMC at file[^\n]+line 6 function \w+\nResuming from jump target 'file[^\n]+line 9 function \w+'\nBMC at file[^\n]+line 9 function \w+(.|\n)+VERIFICATION SUCCESSFUL(.|\n)+BMC at file[^\n]+line 6 function \w+\nResuming from next instruction 'file[^\n]+line 7 function \w+'\nBMC at file[^\n]+line 7 function \w+(.|\n)+Saving next instruction 'file[^\n]+line 7 function \w+'\nSaving jump target 'file[^\n]+line 9 function \w+'(.|\n)+BMC at file[^\n]+line 6 function \w+\nResuming from jump target 'file[^\n]+line 9 function \w+'\nBMC at file[^\n]+line 9 function \w+(.|\n)+VERIFICATION FAILED(.|\n)+BMC at file[^\n]+line 6 function \w+\nResuming from next instruction 'file[^\n]+line 7 function \w+'\nBMC at file[^\n]+line 7 function \w+(.|\n)+VERIFICATION SUCCESSFUL\n[^.]
--
^warning: ignoring
--
This strategy differs from 'fifo' because the failed path is the second
one, not the last one.

WARNING: This test.desc file is automatically generated
from the test.in file in this directory. Do not modify
this file.
```

karkhaz committed:
Symex paths saved onto abstract data structure

This commit introduces a new data type, path_queuet, representing saved
symbolic execution paths, used instead of the std::list. This allows us
to implement custom logic for saving and removing paths, rather than
simply pushing to the back and popping the front.

This commit introduces a concrete instantiation of path_queuet, called
path_fifot, with the same functionality as the old std::list. This is in
preparation for further instantiations that implement more sophisticated
logic for deciding which path should be resumed next.

karkhaz committed:
Bugfix: always print path exploration resume point

Before this commit, CBMC would occasionally print "Resuming from ''"
because the source location that it was accessing was null.

karkhaz committed:
Path exploration pushes both successors onto queue

When doing path exploration and encountering a conditional goto
instruction, both successors of the goto (the next instruction, and the
target) are pushed onto the path exploration workqueue. This allows the
symbolic execution caller to have complete control over which path to
execute next.

Prior to this commit, symex would push one successor onto the workqueue
and continue executing the path of the other successor until the path
was terminated. This commit introduces the possibility that symbolic
execution pauses without reaching the end of a path, and needs to be
resumed by the caller. Thus, new safety checker results are introduced
(SUSPENDED and UNKNOWN); the former signals to the caller of symex that
symex has not reached the end of a branch, and therefore no safety check
has been performed.

The (currently only) path exploration strategy still pops paths in the
order that they were pushed. This behaviour means that the paths that it
pops will always alternate between the next instruction of a goto, and
the target of that same goto. The test suite is updated to reflect this
strategy.

karkhaz commented:
Also, you might suppose that the code formatting is a bit bizarre in the constructor for `path_strategy_choosert` (it contains a literal map from strings to literal pairs of strings and lambda thunks). That's what clang-format told me to do, but I'm happy to change it if anybody else wants it a particular way.

thk123 commented:
TG pointer bump (due to change in bmc.cpp) diffblue/test-gen#1567

karkhaz commented:
@thk123 Not sure what that issue is, as it's private: should I know what this is about, or is it just for diffblue folks?

romainbrenguier commented on src/cbmc/bmc.cpp:
```diff
@@ -596,15 +598,22 @@ void bmct::setup_unwind()
-  safety_checkert::resultt result;
-  goto_symext::branch_worklistt worklist;
+  safety_checkert::resultt final_result = safety_checkert::resultt::UNKNOWN;
+  safety_checkert::resultt tmp_result = safety_checkert::resultt::UNKNOWN;
```
I think `tmp_result` shouldn't be declared here

romainbrenguier commented on src/cbmc/bmc.cpp:
```diff
@@ -614,20 +623,22 @@ int bmct::do_language_agnostic_bmc(
       bmc.set_ui(ui);
       frontend_configure_bmc(bmc, goto_model);
-      result = bmc.run(goto_model.goto_functions);
+      tmp_result = bmc.run(goto_model.goto_functions);
```
`tmp_result` should be const declared here

romainbrenguier commented on src/goto-programs/safety_checker.h:
```diff
@@ -47,11 +60,24 @@ class safety_checkert:public messaget
+    a != safety_checkert::resultt::SUSPENDED &&
+    b != safety_checkert::resultt::SUSPENDED,
+    "Cannot compare safety check result to result of suspended symbolic "
+    "execution");
```
this could be a PRECONDITION

thk123 commented:
@karkhaz Hi sorry should have explained - it is a PR on a private repo to verify this change doesn't require changes on our end to incorporate it. Nothing to worry about, I'll post here when it has passed but I'd appreciate if we could hold off merging until that check is complete :slightly_smiling_face:

thk123 reviewed (dismissed):
Sorry - there are compile errors on our end, blocking until we've had a chance to get this working.

thk123 commented:
Could I ask for a rebase as I'm getting a couple of unrelated compile errors

thk123 commented:
Could I ask for a rebase as I think I'm getting some unrelated compile errors that I think a rebase would fix. (I think I've fixed the real compile errors but would like to check).

karkhaz commented:
@thk123 I have rebased and also applied @romainbrenguier's suggestion.

thk123 commented:
@karkhaz thanks, got it compiling locally, will let you know once our CI is passing.

smowton commented on src/goto-programs/safety_checker.h:
```diff
@@ -30,7 +30,20 @@ class safety_checkert:public messaget
-  enum class resultt { SAFE, UNSAFE, ERROR };
+  enum class resultt
+  {
+    SAFE,
```
Might as well document the existing ones while you're there!

tautschnig replied:
(done)

thk123 commented:
Hi it is failing our CI so until we've investigated can I ask we continue holding off on merging this.

karkhaz commented:
@thk123 please let me know if there's anything I can help with. I'm quite surprised, as I believe that I've only changed aspects of symex that run when `--paths` is passed...

@tautschnig suggested an improvement to the test structure today, so please don't merge this until I've pushed that up (I'll do that tomorrow morning).

@martin-cs would you mind seeing if you're happy with this approach? This patchset implements your suggestion that symex should save both potential paths rather than saving one and continuing with the other.

thk123 commented:
@karkhaz yup looks like entirely our problem, but could I ask for another rebase, sorry.

karkhaz commented:
@thk123 rebased, cheers.

thk123 commented:
@karkhaz thanks - updated the bump :+1:

karkhaz commented:
I've addressed @smowton's suggestion. I've also updated the tests: tests for the different strategies are now all under the `cbmc-path-strategies` directory, with multiple tests per test program; the structure is described in `regression/cbmc-path-strategies/README`.

No idea why the doxygen check is failing.

smowton commented:
Don't get the Doxygen error on my local machine (1.8.13) but do on Travis (1.8.6), so might just be a bug. Possible workaround: setting `EXTENSION_MAPPING = h=c++` in the doxyfile?

karkhaz committed:
Doxygen uses C++ parser for .h files

smowton commented:
Hey all, I've written up what I see as the path exploration technical debt here: https://github.com/diffblue/cbmc/issues/1945

Not saying it needs to happen in this PR, just that it should happen sooner or later.

tautschnig commented on src/goto-symex/path_storage.h:
```diff
@@ -0,0 +1,189 @@
+          {" progressive-fifo             prefer to make progress: all goto\n"
+           "                              targets are popped in fifo order,\n"
+           "                              followed by all next instructions\n",
+           []() {
```
Requires `// NOLINT(whitespace/braces)`.

tautschnig commented on src/goto-symex/path_storage.h:
```diff
@@ -0,0 +1,189 @@
+               new progressive_path_fifot());
+           }}}})
+  {
+  }
```
I'm wondering whether doxygen could be helped by moving this constructor definition to the cpp file. It doesn't seem to be on a hot path, hence no need to place it in the header file anyway.

tautschnig commented on regression/cbmc-path-strategies/if/fifo.desc:
```diff
@@ -0,0 +1,16 @@
+             WARNING:
+This .desc file is automatically generated
+from the file 'if/fifo.in'.
+Do not modify this file.
```
No newline at end-of-file.

tautschnig commented on src/cbmc/bmc.cpp:
```diff
@@ -604,7 +604,8 @@ int bmct::do_language_agnostic_bmc(
   safety_checkert::resultt result;
-  goto_symext::branch_worklistt worklist;
+  std::unique_ptr<path_storaget> worklist =
+    path_storaget::make(path_storaget::disciplinet::FIFO);
```
You might keep the diff small by doing:
````c++
std::unique_ptr<path_storaget> worklist_ptr =
  path_storaget::make(path_storaget::disciplinet::FIFO);
path_storaget &worklist = *worklist_ptr;
````

tautschnig commented on regression/cbmc-path-strategies/Makefile:
```diff
@@ -0,0 +1,23 @@
+	@../test.pl -j 12 -p -c "../../../src/cbmc/cbmc --verbosity 10"
+
+gen-desc:
+	@../../scripts/make_descs.py
```
This rule ought to have dependencies (possible using `$(wildcard ...)` or $(shell find ...)`). Ideally, it would also not be a phony target, but instead explicitly list the files generated (list of files generated using `$(subst ...)` from the list of input files.

tautschnig commented on regression/cbmc-path-strategies/Makefile:
```diff
@@ -0,0 +1,23 @@
+	done;
+
+clean:
+	find -name 'test.desc' -execdir $(RM) '{}' \;
```
That's not quite complete, is it?

tautschnig commented on scripts/make_descs.py:
```diff
@@ -0,0 +1,180 @@
+                    "             WARNING:",
+                    "This .desc file is automatically generated",
+                    "from the file '{fyle}'.".format(fyle=in_file),
+                    "Do not modify this file."
```
This needs a `\n` at the end to get rid of my complaint about missing-newline-at-end-of-file.

tautschnig commented on regression/cbmc-path-strategies/Makefile:
```diff
@@ -1,10 +1,10 @@
 
 tests.log: ../test.pl gen-desc
-	@../test.pl -j 12 -p -c "../../../src/cbmc/cbmc --verbosity 10"
+	@../test.pl -p -c "../../../src/cbmc/cbmc --verbosity 10"
```
Merge this change into the prior commit.

karkhaz commented:
All checks have gone through after I moved the `path_strategy_choosert` constructor to the `.cpp` file, minus the usual clang-format complaints about `{c,j}bmc_parse_optionst` constructor.

@tautschnig thanks for pointing out the issue with the `.desc` files. Those files are generated and should not have been committed at all! I've removed them, and also ensured that the generator script adds a newline at the end of the generated files.

karkhaz commented:
Just rebased this on top of whichever PR was causing conflicts, it should be good to go.

smowton commented on .gitignore:
```diff
@@ -57,6 +57,9 @@ regression/**/*.smt2
 /regression/coverage_**
 
+# These files are generated from the test.in file in their directories
+regression/cbmc-path-strategies/*/*.desc
```
Nitpick (which obviously hasn't been followed above): usually you would put something like this in the `regression/cbmc-path-strategies` directory, whereupon the `*/` is no longer needed

smowton commented on scripts/make_descs.py:
```diff
@@ -0,0 +1,193 @@
+#!/usr/bin/env python3
```
Nit: document and check for your minimum required version of py3, if known

smowton commented on scripts/make_descs.py:
```diff
@@ -0,0 +1,193 @@
+        for line in f:
+            cnt += 1
+            line = line.strip()
+            if re.match("^--$", line):
```
surely `line == "--"`?

smowton commented on src/cbmc/bmc.cpp:
```diff
@@ -377,6 +377,9 @@ safety_checkert::resultt bmct::execute(
     const goto_functionst &goto_functions =
       goto_model.get_goto_functions();
 
+    if(symex.states_pushed)
```
Add a comment briefly explaining what condition we've spotted here?

smowton commented on src/goto-symex/goto_symex.h:
```diff
@@ -177,6 +160,15 @@ class goto_symext
+  /// doing path exploration, and thus pushed the successor states of the GOTO
+  /// onto path_storage. The symbolic execution caller should now choose which
+  /// successor state to continue executing, and resume symex from that state.
+  bool states_pushed;
```
Given this description, should we just name the variable `should_suspend_symex`? That would address my comment request above too.

smowton commented on src/goto-symex/path_storage.cpp:
```diff
@@ -0,0 +1,143 @@
+         "                              goto target; paths are popped in\n"
+         "                              first-in, first-out order\n",
+         []() { // NOLINT(whitespace/braces)
+           return std::unique_ptr<path_fifot>(new path_fifot());
```
Use `util/make_unique.h`

smowton reviewed (commented):
Looks ok for the most part, my only concern is the test mechanism, which strikes me as a bit obscure, since it adds a custom scripting language that generates our own peculiar test-description format... would this stuff be better expressed as a unit test, which could directly check for the expected operation sequence instead of compiling a massive regex that parses CBMC's debug output? Thoughts @tautschnig @peterschrammel ?

smowton commented:
If unit tests were used, this would also turn the path "custom scripting language -> Python compiler -> test.desc -> perl script -> result" into "test description -> c++ verification code -> result", which sounds easier to debug.

karkhaz commented:
@smowton thanks for the suggestion. If others are happy with this, I can write unit tests instead. I won't do this straight away, as I've not written unit tests for CBMC before, but will try to get round to it soon.

tautschnig commented:
FWIW and also in response to @smowton's question: @karkhaz and I discussion out-of-band that unit tests should be the way to go here.

karkhaz commented:
@smowton @tautschnig I've just read through some of the existing tests under `unit/`, and I'm starting to disagree that unit tests are the correct way to test for this. I can't see a way of testing the correct order of pushes and pops unless we have a fully-fledged goto-program to test with (in some cases, including functions and function calls, etc). I would need to artificially construct a goto-program for each test case programatically, whereas `regression` allows me to write the input as a `.c` file. Are there any unit tests that do read in a `.c` file---I couldn't find any?

If you have any pointers to unit tests that do something similar to what I need, I'd appreciate it. But otherwise I think that we do need to use an end-to-end test, because there's way too much setup needed before calling symex that I would need to do programatically. I also suspect that programatically replicating the setup that the frontends, BMC, etc. do would be a maintenance burden every time something changes in that code. Please let me know if you think otherwise.

tautschnig commented:
@karkhaz You might take a look at https://github.com/tautschnig/fshell/blob/master/fshell2/tc_generation/constraint_strengthening.t.cpp (and a few other .t.cpp files in that project). I don't think we have any such tests in CBMC yet, but this might provide an idea on how it could be done? Arguably this is against CBMC 5.6, so there are a number of changes to various APIs and BMC set-up might look a bit different now.

karkhaz committed:
Make new LIFO path exploration the default

This commits adds the "lifo" strategy, which explores program paths
using depth-first search. This is now the default strategy due to its
favourable memory consumption compared to other strategies; tests on a
few benchmarks show that its memory usage is basically constant and
comparable to regular model-checking, since this strategy finishes paths
as quickly as possible rather than saving them for later.

karkhaz commented:
@smowton I've addressed all your other comments, thanks. I've also added an additional commit: the new `lifo` strategy, which explores programs using a depth-first search, is now the default strategy due to it having constant memory usage (as opposed to the other two strategies, which are closer to a breadth-first search).

@tautschnig thank you, I'll have a try at this.

smowton commented:
@karkhaz here's an example of a unit test that creates a GOTO program from Java source, would that work as a template for you? https://github.com/diffblue/cbmc/blob/develop/unit/java_bytecode/java_virtual_functions/virtual_functions.cpp

Note line 47, which loads a Java class file, and then line 58 which converts it to GOTO. You'd need to then run `remove_returns` and other expected passes (i.e. relevant passes from https://github.com/diffblue/cbmc/blob/develop/src/cbmc/cbmc_parse_options.cpp#L711), before making a `bmct` and testing its results.

Making an equivalent of `load_java_class` that instantiates the C frontend would probably not be too hard if for some reason the GOTO that results from Java is inappropriate for your needs.

karkhaz commented:
@smowton thanks for the suggestion, I already started work on the unit test though, I think the infrastructure is fairly complete---I just need to translate my old regression tests over.

@smowton @tautschnig could I ask for your opinion on how this unit test is written? It's in the file [unit/path_strategies.cpp](https://github.com/diffblue/cbmc/pull/1915/files#diff-e2e02bccddd85df1141068eef30f2212) in the commit that I just pushed. There's a fair bit of infrastructure, but I'm hoping that the 'interface' that people specify the tests with is fairly simple. For each test program, there are a bunch of literal lists of symex "events", one list for each strategy. Events are either peeks at the worklist (to check that we're peeking at paths in the correct order), or results (for when we reach the end of a path and model-check).

karkhaz commented:
Oh, and since I pushed a WIP, please could somebody add the do-not-merge label to this PR---thank you.

karkhaz commented:
OK, having implemented this properly, IMO this seems like it's duplicating a vast amount of functionality from `cbmc_parse_optionst` and `bmct`. It's looking like a regression test in the guise of a unit test, and I don't think it's going to be more maintainable than what I had before. I'd welcome suggestions on how to improve it, though.

Specifically, I'd like to write down the sequence of peeked paths and verification results that I'd expect for each path strategy. However, in order to match the line numbers with the ones output by `cbmc --show-goto-functions`, I need to do a tonne of preprocessing that `cbmc_parse_optionst` does (removing skips, etc.). So I made `cbmc_parse_optionst::get_goto_model` into a `static` method (as well as everything that it calls) so that I could call it from my test harness. I also need to set all of the options that `cbmc_parse_optionst` sets to `true` by default (`assertions`, `propagation`, etc.). So there's a lot of duplication there too. Then, I had to write a simplified version of `bmct::do_language_agnostic_bmc` so that I could inspect the worklist every time symex returned.

Additionally, for some reason that I've spent the last day failing to figure out, some of the verification results don't match the ones when I run `cbmc`. On this test case:
```
  GIVEN("a simple loop program")
  {
    c =
      "/*  1 */   int main()                       \n"
      "/*  2 */   {                                \n"
      "/*  3 */     int x;                         \n"
      "/*  4 */     __CPROVER_assume(x == 1);      \n"
      "/*  5 */                                    \n"
      "/*  6 */     while(x)                       \n"
      "/*  7 */       --x;                         \n"
      "/*  8 */                                    \n"
      "/*  9 */     assert(x);                     \n"
      "/* 10 */   }                                \n";
    const unsigned unwind_limit = 2U;

    check_with_strategy(
      "lifo",
      c,
      {// The path where we skip the loop body. Successful because the path is
       // implausible, we cannot have skipped the body if x == 1.
       symex_eventt::resume(symex_eventt::enumt::JUMP, 9),
       symex_eventt::result(symex_eventt::enumt::SUCCESS),

       // Enter the loop body once. Since we decrement x, the assertion should
       // fail.
       symex_eventt::resume(symex_eventt::enumt::NEXT, 7),
       symex_eventt::resume(symex_eventt::enumt::JUMP, 9),
       symex_eventt::result(symex_eventt::enumt::FAILURE),  // THIS PATH
                                                            // SUCCEEDS INSTEAD
                                                            // OF FAILING

       // The path where we enter the loop body twice. Successful because
       // infeasible.
       symex_eventt::resume(symex_eventt::enumt::NEXT, 7),
       symex_eventt::result(symex_eventt::enumt::SUCCESS),
       },
      unwind_limit);
```
On the event that I commented on, I expect the assertion to fail because we've been through the loop once (decrementing `x` to 0) and then doing `assert(x)`. For some reason, the model checker doesn't pick this up at all. I'm sure that I've set all the same options that are set in CBMC (I dumped the `options_map` from `cbmc_parse_optionst` to make sure, and then replicated it in my test harness), but for some reason the solver just doesn't notice the assertion.

Thoughts about this? Even if I manage to figure out why that test case is failing, I suggest that we go with the original plan of matching on debug output. It's a bit scruffier, but this way seems to necessitate a tonne of duplicated code as far as I can see.

tautschnig commented:
On debugging first: I'd suggest to use `__CPROVER_assert(x, "")` instead of `assert`.  Assuming that by itself doesn't fix the problem, I'd dump the sequence of goto program instructions seen during this run and, more importantly, the resulting `symex_target_equationt`.

Now with regard  to unit tests with lots of code duplication vs. regression tests. I like both approaches: the latter is more succinct (if just ignoring the extra python script), while the former provides a lot of additional documentation. It documents a lot of implicit assumptions. That does mean a lot of debugging pain at times (see above), but might be an exercise with future value?

karkhaz commented:
`__CPROVER_assert(x, "")` doesn't work, in fact neither does `__CPROVER_assert(0, "")`. But that line is executed. Here's the debug output you asked for (with `__CPROVER_assert(0, "")`), for the final two paths on that test (one should fail and the second should pass). I'm not sure how to interpret the equation though.

*Edit* by "doesn't work", I mean verification is successful, which is not what I expect.

https://pastebin.com/F6zPbZiv

About unit tests: I agree that this exercise has highlighted lots of implicit assumptions. Therefore if I'm going to go with unit tests, I propose that I go all the way with it, and move all of the implicit logic of `cbmc_parse_optionst` into static functions that can be called by anybody. For example, there's a lot of this sort of thing:
```
if(cmdline.isset("no-propagation"))
    options.set_option("propagation", false);
  else
    options.set_option("propagation", true);
```
I would rather have everything that defaults to `true` set in one place (a static method) that both `cbmc_parse_optionst` and my unit test can call into, and then `cbmc_parse_optionst` can do the `if` cases but not bother with the `else`s. Then, anybody wanting to implement a default would be encouraged to put it in the static method, so that it gets picked up by unit tests also and be useful to e.g. #1962. What do you think about this approach?

karkhaz commented:
(just pushed the code that I have at the moment).

tautschnig commented:
The pasted debug log looks fine, what are the output log lines for the SAT solver (solving using ..., x variables, y clauses, formula is unsatisfiable)?

On the implementation plan:  it would be great to have all those defaults factored out. @kroening  has made some initial effort in `goto-programs/initialize_goto_model.{h,cpp}`, and likely there should be a lot more such static functions as you propose. So ... if you can ... that would be very much appreciated!

smowton commented:
Re: `get_goto_program`: do we really need most of the get-goto-program passes for this case, e.g. remove_function_pointers etc? For these simple input programs I imagine it's not necessary.

Similarly how sure are you you need all those particular command-line options?

smowton commented:
The strategy looks good to me btw -- it is a bit verbose, but this way it's far easier to debug when something goes awry, compared to the big-regex approach

karkhaz commented:
Never mind, I messed up, my fault. I had passed `[](){ return true; }` as the symex callback, which makes `bmct` avoid running the solver. I should have returned false.

I'll complete the test migration. @smowton, I don't think all of the goto-program passes are necessary (removing skips definitely is though, as sometimes a goto will point to the skip before the instruction that I think it should point to). I feel that just calling into `cbmc_parse_optionst` to get a goto-model the same way that CBMC itself does is more maintainable though, rather than having to keep the unit test in sync with how CBMC does things.

karkhaz committed:
Unit tests for path exploration strategies

karkhaz commented:
Hi all, I've completed the unit test, removed the old regression tests, and done some refactoring of `cbmc_parse_optionst` to support the unit test.

This will fail to build using the Makefile although it's building using CMake, could I have some advice about this? I modified [`unit/CMakeLists.txt`](https://github.com/diffblue/cbmc/pull/1915/files#diff-b4ab751cb778d42993621801a4c26e3a), adding `cbmc-lib` to the list of `target_link_libraries`, since the unit test needs to link to `bmct` and other things. However, I cannot figure out what the equivalent of this in [`unit/Makefile`](https://github.com/diffblue/cbmc/pull/1915/files#diff-7e395b52a0308dba5e705fe157750433) would be. I tried adding `../src/cbmc/cbmc$(LIBEXT)` to the list of `CPROVER_LIBS`, but that doesn't work.

tautschnig commented:
> I tried adding `../src/cbmc/cbmc$(LIBEXT)` to the list of `CPROVER_LIBS`, but that doesn't work.

The `cbmc` folder does not currently build a library (when using Makefiles). You might instead  just refer to the required `xyz$(OBJEXT)` files instead.

karkhaz committed:
Add path strategy chooser class

A new class is introduced to hold the names, descriptions and
constructor thunks for all path exploration strategies. This provides a
uniform way for strategy descriptions to be displayed, which can be done
with the --show-symex-strategies flag.

karkhaz commented:
I've amended the `unit` Makefile to link against all necessary object files; all tests are passing. Thanks!

tautschnig commented:
Given the size of the PR and to help speed up things: could you please create another PR that just has the two "Factor out ..." commits?

karkhaz commented:
Sure, although of course the checks for this PR will not pass until the "Factor out" commits have merged, since "Unit tests..." depends on those two commits. Should I leave the "Unit tests..." commit in this PR, or put it in a third PR by itself?

tautschnig commented:
I'd suggest to leave all the commits in this PR as they are, ask reviewers to review commit-by-commit, and point out that the two "Factor out ..." exist elsewhere. Once that new PR is merged the rebase ought to be trivial, so I'm hoping not to generate serious extra work for you.

karkhaz commented:
Yes, the rebase will be easy, no worries. Opened #2049. Thanks!

thk123 commented on unit/path_strategies.cpp:
```diff
@@ -0,0 +1,374 @@
+
+  register_language(new_ansi_c_language);
+  cmdlinet cmdline;
+  cmdline.set("function", "main");
```
Watch out - here if you use the `cmdlinet` the `set` call is ignored as it doesn't think `function` etc is a real flag. I wrote `free_form_cmdlinet` that automatically adds the flags as you set them.

karkhaz replied:
@thk123 I don't understand this comment, if I remove that line the test no longer works (it doesn't automatically set the entry point to `main`), so this call must be having some effect?

thk123 replied:
That's surprising, not least because `main` is the default entry point! If easy for you to do, I suggest stepping through `cmdlinet::set` cos I found that it doesn't normally do anything.

tautschnig replied:
From reading the code in `cmdline.cpp` I would agree with @thk123, so something weird seems to be happening here. I note that the next line, `cmdline.set("paths", strategy);` is likely redundant as that's all taken care of via `opts.set_option` calls later on. Furthermore it seems that `unit/pointer-analysis/custom_value_set_analysis.cpp` also (successfully?) uses `cmdlinet` in the way it is done here. Insights would be really appreciated!

karkhaz replied:
If I remove either `cmdline.set("function", "main");` or `config.main = "main";`, I get
```
Type-checking path-explore_18497.A3WUtT
1: file /tmp/path-explore_18497.A3WUtT.c line 9 function main: function `assert' is not declared
1: main symbol `E.f' not found
```
And everything fails. Do you have any idea why this happens, or should I step through or print out more debug info?

tautschnig replied:
`config.set(cmdline);`  should take care of setting `config.main` if the "function" option was set in `cmdline`. So `config.main = "main";` being required is consistent with `cmdline.set` not having any effect. But `cmdline.set("function", "main");` being vital is still a mystery.

karkhaz replied:
:sob: of course, it's because I'm slightly distracted at a conference, and so I'm running tests locally on my laptop but modifying the test suite through SSH to my workstation.

I've removed the `set` calls (including the `"paths"` option), thanks for the review. This PR needs rebasing since #2049 was merged, but I'll do that when I'm less distracted...

karkhaz commented:
I've rebased this to tip-of-tree, it no longer includes the commits merged in #2049. All tests are passing.

karkhaz commented:
Ping @smowton, @thk123---does this look good to everyone? I'm experimenting with new strategies locally, but I don't want to get too out-of-sync if there are many more changes required from this PR...

smowton commented on unit/Makefile:
```diff
@@ -66,6 +67,34 @@ cprover.dir:
+          ../src/cbmc/symex_bmc$(OBJEXT) \
+          ../src/cbmc/symex_coverage$(OBJEXT) \
+          ../src/cbmc/xml_interface$(OBJEXT) \
+					../src/jsil/expr2jsil$(OBJEXT) \
```
Alignment looks odd --mix of spaces and tabs?

thk123 commented:
Updating the pointer bump now - note I've not really reviewed the PR

smowton reviewed (approved):
Re-reviewed the tests -- this looks excellent, thanks very much for the extra work! Only one nitpick, a Makefile has some funky tab/space combo at the moment. No need for re-review from me.

thk123 commented:
Sorry to create extra work - but could I ask for a rebase as this has fallen behind develop.

karkhaz commented:
@smowton fixed the nit, thanks!

@thk123 rebased to ToT.

All tests are passing.

thk123 commented:
Apologies - we had a number of breaking changes this week that got merged after your last rebase. Can I ask for _another_ rebase (I'm optimistic it will be the last!)

karkhaz commented:
@thk123 rebased, cheers.

thk123 reviewed (approved):
:tada: TG pointer bump passes ~@karkhaz I assume you'd like this merged post-haste?~ still waiting for approval from I honestly don't know who - something that even @tautschnig or @smowton don't have permission to wave through?

tautschnig commented:
@chrisr-diffblue @martin-cs @peterschrammel A review would be greatly appreciated!

chrisr-diffblue commented on src/doxyfile:
```diff
@@ -281,7 +281,7 @@ OPTIMIZE_OUTPUT_VHDL   = NO
 # the files are not read by doxygen.
 
-EXTENSION_MAPPING      =
+EXTENSION_MAPPING      = h=c++
```
Minor nit: This should probably have been a separate PR... but I won't block the PR :-)

chrisr-diffblue reviewed (approved):
Looks good to me - many thanks for all the efforts to get this through PR.

karkhaz commented:
@chrisr-diffblue I'll put minor commits in a separate PR from now on, IIRC I kept that commit in because the PR was failing the doxygen check that that commit fixes.

Thanks so much for all the reviews!
